src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php

```diff
@@ -2349,11 +2352,38 @@ private function registerHttpClientConfiguration(array $config, ContainerBuilder
             $container->removeDefinition(HttpClient::class);
         }
 
-        if ($this->readConfigEnabled('http_client.retry_failed', $container, $retryOptions)) {
+        if ($hasRetryFailed = $this->readConfigEnabled('http_client.retry_failed', $container, $retryOptions)) {
             $this->registerRetryableHttpClient($retryOptions, 'http_client', $container);
         }
 
-        $httpClientId = ($retryOptions['enabled'] ?? false) ? 'http_client.retryable.inner' : ($this->isInitializedConfigEnabled('profiler') ? '.debug.http_client.inner' : 'http_client');
+        if ($hasUriTemplate = class_exists(UriTemplateHttpClient::class)) {
+            $uriTemplateExpanderId = null;
+            if (ContainerBuilder::willBeAvailable('guzzlehttp/uri-template', \GuzzleHttp\UriTemplate\UriTemplate::class, [])) {
+                $uriTemplateExpanderId = 'http_client.uri_template_expander.guzzle';
+            } elseif (ContainerBuilder::willBeAvailable('rize/uri-template', \Rize\UriTemplate::class, [])) {
+                $uriTemplateExpanderId = 'http_client.uri_template_expander.rize';
+            }
+
+            if ($uriTemplateExpanderId) {
+                $container->setAlias('http_client.uri_template_expander', $uriTemplateExpanderId);
+            }
+
+            $container
+                ->getDefinition('http_client.uri_template')
+                ->setArgument(2, $defaultUriTemplateVars);
+        } elseif ($defaultUriTemplateVars) {
+            throw new LogicException('Support for URI template requires symfony/http-client 6.3 or higher, try upgrading.');
+        }
+
+        $httpClientId = 'http_client';
+        if ($hasUriTemplate) {
+            $httpClientId = 'http_client.uri_template.inner';
+        } elseif ($hasRetryFailed) {
+            $httpClientId = 'http_client.retryable.inner';
+        } elseif ($this->isInitializedConfigEnabled('profiler')) {
+            $httpClientId = '.debug.http_client.inner';
```

Comment:
This does not make sense to me. the MockClient should be configured as a decorator service with the right priority instead of trying to access the inner service of another decorator.

Reply:
and if scoped client need to inject the inner HttpClient before any decoration is applied on them, a better solution would be to register that raw client using an internal id (that is never decorated in the config) and then have `http_client` being an alias of it (which is then decorated as needed) instead of this unmaintainable list.

Reply:
:exploding_head: It seems to work but isn't it a BC break to change `http_client` from a definition to an alias?

Reply:
I would say now: as soon as decoarators are resolved, it will be an alias anyway.

Reply:
I've opened https://github.com/symfony/symfony/pull/49513. In this PR, I'll keep things as they are.